python/mp/rosary.py

```diff
@@ -354,14 +356,27 @@ def mainloop(self):
                 if (effect.finished):
                     self.del_effect(effect.id)
 
-            self.update()
-
             # Let the triggers figure out for themselves what to do
             for trigger in self.triggers.values():
                 if trigger.running:
                     trigger.next()
 
-            time.sleep(self.mainloop_delay)
+            # drop a frame if we've already passed next_frame_time
+            if (time.monotonic() > next_frame_time):
+                print('frame drop')
+                next_frame_time += self.frame_time
+
+                # "dropping a frame" means calling next() on all the effects w/o updating the LEDs
+                for effect in self.effects:
+                    effect.next(self)
+                    if (effect.finished):
+                        self.del_effect(effect.id)
+
+            next_frame_time += self.frame_time
+
+            self.update()
+
+            time.sleep(next_frame_time - time.monotonic())
 
```

Comment:
This is my fixed-frame rate change, which I told you to merge in. Let's just keep it here and merge it.